pact/hft.pact

```diff
@@ -369,6 +394,136 @@
+    @doc "Wrapper cap/event of SALE of TOKEN by SELLER of AMOUNT until TIMEOUT block height."
+    @event
+    (compose-capability (OFFER token seller amount timeout))
+    (with-read tokens token
+      { 'policy := policy:module{token-policy-v1}
+      , 'supply := supply
+      , 'minimum-precision := precision
+      , 'manifest := manifest
+      }
+      (policy::init-sale
+        { 'token: token, 'supply: supply, 'precision: precision, 'manifest: manifest }
+        seller amount sale))
+    (compose-capability (SALE_PRIVATE sale))
+  )
+
+  (defcap OFFER
+    (token:string seller:string amount:decimal timeout:integer)
+    @doc "Managed cap for SELLER offering AMOUNT of TOKEN until TIMEOUT."
+    @managed
+    (enforce (sale-active timeout) "SALE: invalid timeout")
+    (compose-capability (DEBIT token seller))
+    (compose-capability (CREDIT token (sale-account)))
+  )
+
+  (defcap WITHDRAW
+    (token:string seller:string amount:decimal timeout:integer sale:string)
+    @doc "Withdraws offer SALE from SELLER of AMOUNT of TOKEN after timeout."
+    @event
+    (enforce (not (sale-active timeout)) "WITHDRAW: still active")
+    (compose-capability (DEBIT token (sale-account)))
+    (compose-capability (CREDIT token seller))
+    (compose-capability (SALE_PRIVATE sale))
+  )
+
+  (defcap BUY
+    (token:string seller:string buyer:string amount:decimal timeout:integer sale:string)
+    @doc "Completes sale OFFER to BUYER."
+    @managed
+    (enforce (sale-active timeout) "BUY: expired")
+    (with-read tokens token
+      { 'policy := policy:module{token-policy-v1}
+      , 'supply := supply
+      , 'minimum-precision := precision
+      , 'manifest := manifest
+      }
+      (policy::enforce-sale
+        { 'token: token, 'supply: supply, 'precision: precision, 'manifest: manifest }
+        seller buyer amount sale))
+    (compose-capability (DEBIT token (sale-account)))
+    (compose-capability (CREDIT token buyer))
+    (compose-capability (SALE_PRIVATE sale))
+  )
+
+  (defcap SALE_PRIVATE (sale:string) true)
+
+
+  (defpact sale
+    ( token:string
+      seller:string
+      amount:decimal
+      timeout:integer
+    )
+    (step-with-rollback
+      (with-capability (SALE token seller amount timeout (pact-id))
+        (offer token seller amount timeout))
+      (with-capability (WITHDRAW token seller amount timeout (pact-id))
+        (withdraw token seller amount))
+    )
+    (step
+      (let ( (buyer:string (read-msg "buyer"))
+             (buyer-guard:guard (read-msg "buyer-guard")) )
+        (with-capability (BUY token seller buyer amount timeout (pact-id))
+          (buy token seller buyer buyer-guard amount))))
+  )
+
+  (defun offer
+    ( token:string
+      seller:string
+      amount:decimal
+    )
+    @doc "Initiate sale with by SELLER by escrowing AMOUNT of TOKEN until TIMEOUT."
+    (require-capability (SALE_PRIVATE (pact-id)))
+    (debit token seller amount)
```

Comment:
Looks like the capability `DEBIT` is not granted to this function. I also tried adding `(with-capability (OFFER ..))` to wrap around `debit` and `credit` but still doesn't resolve.